packages/hardhat-kadena/contracts/Redeem.yul

```diff
@@ -0,0 +1,177 @@
+    let sigR := calldataload(0x9b)
+    let sigS := calldataload(0xbb)
+    let rootTime := shr(0xc0, calldataload(0xdb))
+
+    /* ********************************************************************** */
+    /* Check Version */
+
+    if iszero(eq(version, 0x0)) {
+      error(0x02)
+    }
+
+    /* ********************************************************************** */
+    /* Check Target ChainId */
+
+    let chainwebChainIdContract := 0x9b02c3e2df42533e0fd166798b5a616f59dbd2cc
+    if iszero(staticcall(gas(), chainwebChainIdContract, 0, 0, add(chainIdPtr, 0x1c), 0x4)) {
+      error(0x03)
+    }
+    if iszero(eq(mload(chainIdPtr), targetChain)) {
+      error(0x04)
+    }
+
+    /* ********************************************************************** */
+    /* Compute Message Hash */
+
+    calldatacopy(msgPtr, msgOffset, msgSize)
+    mstore(msgHashPtr, keccak256(msgPtr, msgSize))
+
+    /* ********************************************************************** */
+    /* Check Signature */
+
+    // copy v value (right aligned)
+    calldatacopy(add(sigPtr, 0x1f), sigOffset, 0x01)
+    // copy r and s value
+    calldatacopy(add(sigPtr, 0x20), add(sigOffset, 0x01), 0x40)
+
+    let ecrecoverContract := 0x01
+    if iszero(staticcall(gas(), ecrecoverContract, msgHashPtr, 0x80, sigAddrPtr, 0x20)) {
+      error(0x05)
+    }
+    if iszero(eq(mload(sigAddrPtr), REDEEM_KEY_ADDR)) {
+      error(0x06)
+    }
+
+    /* ********************************************************************** */
+    /* Query Header Oracle */
+
+    // Use EIP-4788 to check whether a header is included on chain.  For that
+    // the timestamp of the header is needed. This information is merly
+    // informational and does not need to be verfied.
+    //
+    // Also note that only the most recent 8191 headers are available which is
+    // about 3 days on the Kadena mainnet.
+
+    let beaconRootsContract := 0x000F3df6D732807Ef1319fB7B8bB8522d0Beac02
+    mstore(rootTimePtr, rootTime)
+    if iszero(staticcall(gas(), beaconRootsContract, rootTimePtr, 0x20, rootPtr, 0x20)) {
+      error(0x07)
+    }
+    if iszero(eq(mload(rootPtr), rootHash)) {
+      error(0x08)
+    }
+
+    /* ********************************************************************** */
+    /* Check Amount */
+
+    let redeemedAmount := sload(xChanId)
+    let maxAmount := sub(xChanBalance, redeemedAmount)
+
+    if gt(amount, maxAmount) {
+      error(0x09)
+    }
+
+    /* ********************************************************************** */
+    /* Update Redeemed Amount */
+
+    sstore(xChanId, add(redeemedAmount, amount))
+
+    /* ********************************************************************** */
+    /* Send Funds to Target Account */
+
+    // TODO: is this safe or do we need to check first that there is no code
+    // at the target account?
+    if iszero(call(gas(), targetAccount, amount, 0, 0, 0, 0)) {
+      error(0x0a)
+    }
```

Comment:
I would check that the address is not a contract. Otherwise, a reentrancy attack could be executed by the contract if it is one.

One disadvantage is that it wouldn't be possible to send native token to a contract (which could be legitimate), but I think the risk outways the advantage. The receiving EOA can send native token to a contract if it wants to.

If you do decide to allow sending native token to a contract, do not send all remaing gas. That would allow the contract to execute code rather than just receive native token. Send a fixed gas stipend. 2300 is the standard fixed stipend for native token transfers.

for static calls (read only), using gas() is fine.

Reply:
if iszero(iszero(extcodesize(targetAccount))) {
  error(0x0b) // or another error code for "target is a contract"
}